pilot/pkg/model/network_gw_names.go

```diff
@@ -0,0 +1,128 @@
+package model
+
+import (
+	"math"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/miekg/dns"
+)
+
+type nameCacheEntry struct {
+	value  []string
+	expiry time.Time
+}
+
+type networkGatewayNameCache struct {
+	sync.Mutex
+	client     *dnsClient
+	cache      map[string]nameCacheEntry
+	xdsUpdater XDSUpdater
+}
+
+func newNetworkGatewayNameCache(xdsUpdater XDSUpdater) (*networkGatewayNameCache, error) {
+	c, err := newClient()
+	if err != nil {
+		return nil, err
+	}
+	return &networkGatewayNameCache{client: c, xdsUpdater: xdsUpdater, cache: map[string]nameCacheEntry{}}, nil
+}
+
+func (n *networkGatewayNameCache) Resolve(name string) []string {
+	n.Lock()
+	defer n.Unlock()
+	if entry, ok := n.cache[name]; ok && entry.expiry.After(time.Now()) {
+		return entry.value
+	}
+	return n.resolveAndCache(name)
+}
+
+func (n *networkGatewayNameCache) resolveAndCache(name string) []string {
+	delete(n.cache, name)
+	names, ttl := n.resolve(name)
+	// don't cache for very small TTL or empty list of names
```

Comment:
For very small TTL, maybe we should just set a minimum bar we cap at. Like never resolve more than 1x/minute?

I don't think any cloud is expecting that all clients are re-resolving DNS every single query and will break if they don't - if they change IP I would hope they would have both IPs working for some period of time. Just a guess - i do not know for sure.